File: eval/fixtures/py-sqli/safe_twin.py
# eval/fixtures/py-sqli/safe_twin.py
# BACO Eval Fixture: Secure SQL twin (parameterized query)
# This is the SECURE version - any finding here is a false positive

from typing import Optional

def get_user_by_id(user_id: int) -> Optional[dict]:
    """
    Fetch user from database by ID.
    
    SECURE: Uses parameterized query to prevent SQL injection.
    The parameter is passed separately, preventing any injection.
    """
    query = "SELECT * FROM users WHERE id = %s"
    params = (user_id,)
    
    # Execute the query with parameters (mock implementation)
    print(f"Executing: {query} with params: {params}")
    
    return None


def delete_user(user_id: int) -> bool:
    """
    Delete a user by ID.
    
    SECURE: Uses parameterized query.
    """
    query = "DELETE FROM users WHERE id = %s"
    params = (user_id,)
    print(f"Executing: {query} with params: {params}")
    return True

File: eval/fixtures/php-sqli/safe_twin.php
<?php
// eval/fixtures/php-sqli/safe_twin.php
// BACO Eval Fixture: SQL Injection safe twin - prepared statement (CWE-89)

require_once 'db.php';

/**
 * Look up a user by the id request parameter.
 *
 * SECURE: uses a prepared statement with a bound parameter, so the
 * user-supplied value can never alter the SQL structure.
 */
function find_user($connection) {
    $user_id = $_GET['id'];
    $stmt = mysqli_prepare($connection, "SELECT * FROM users WHERE id = ?");
    mysqli_stmt_bind_param($stmt, "s", $user_id);
    mysqli_stmt_execute($stmt);
    $result = mysqli_stmt_get_result($stmt);
    return mysqli_fetch_assoc($result);
}
